Torchdrug integration

jannisborn opened the pull request:
TorchDrug integration (inference-only) as discussed @drugilsberg!

- 2 models implemented `GCPN` (Graph Convolutional Policy Network, NeurIPS 2018) and `GAF` (Graph AutoregressiveFlow, ICLR 2020).
- For each models, 3 pretrained models are made available: 
  - zinc250k: `TorchDrugZincGCPN` and `TorchDrugZincGAF`
  - optimized to generated high-qed molecules: `TorchDrugQedGAF` and `TorchDrugQedGCPN`.
  - optimized to generated high-plogp molecules: `TorchDrugPlogpGCPN` and `TorchDrugPlogpGAF`

- unittests implemented for all models

ToDo:
- [x] fix CI pipeline
- [x] upload models on COS (2 times)


## Regarding CI: 

I knew this would backfire, the tests are still failing at installation. Locally, I had to install `pytorch-scatter` via conda, I did the following:
```sh
pip3 install torch==1.8.1 torchvision==0.9.1 torchaudio==0.8.1 -f https://download.pytorch.org/whl/lts/1.8/torch_lts.html
conda install pytorch-scatter -c pyg
pip install torchdrug
```
and this worked fine.  

When I reproduce the CI pipeline locally, I also do not get a functioning installation. The env is created successfully but importing `pytorch-scatter` results in segmentation faults.

Possible solution: Adapt the `conda.yml` like so:
```yml
name: gt4sd
channels:
  - https://conda.anaconda.org/pyg
dependencies:
  - python>=3.7,<3.8
  - pip>=19.1,<20.3
  - pytorch-scatter
  - pip:
      - -r requirements.txt
      - -r vcs_requirements.txt
      # development
      - -r dev_requirements.txt
```

jannisborn commented:
Open for review and discussion about pytorch-scatter/torchdrug installation workflow @drugilsberg!

jannisborn commented:
Final updates on this PR @drugilsberg:
- I parametrized `num_sample` for the generation, as we discussed. I checked it locally, the sample generation is faster now
- openmp is enabled again after the torchdrug imports and it does not interfer with torchdrug-related inference, even if we mix it with inference of other models
- `pytorch-scatter` installation: the package installs stable on ubuntu if we install pytorch via conda too. I changed the `conda.yml`, and the installations runs through now, no more segmentation faults on both mac and ubuntu 🚀  The conda incubator setup goes up from 4 to 10min but I guess that's the price we have to pay for torchdrug 😐 
- I found (and fixed) a terrible bug caused by `torchdrug`. Essentially, they overwrite the default `nn.Module` of torch. See [here](https://github.com/DeepGraphLearning/torchdrug/blob/eeee19181572ef5b8a806b71bdd4d2d1a4e27f67/torchdrug/patch.py#L125). The crazy thing is that **all** subsequent code running native torch uses this patched module. I found it out because they forgot to implement a keyword argument that is present in native torch and that was used by the EnzymeOptimzer (Roberta). I patched it in `torchdrug.implementation`, see bda7951d96ec9573761d2d8999155969d57f5df3 but overall the torchdrug integration raises some concerns regarding the overall code reliability 😐  I opened an issue in the torchdrug repo, but I think we can merge this anyways.

drugilsberg commented on conda.yml:
```diff
@@ -1,9 +1,15 @@
       - -r vcs_requirements.txt
       # development
       - -r dev_requirements.txt
+  - pytorch-scatter
```
this should probably be removed

jannisborn replied:
Done!

drugilsberg commented on conda.yml:
```diff
@@ -1,9 +1,15 @@
 dependencies:
   - python>=3.7,<3.8
   - pip>=19.1,<20.3
+  - pytorch
```
not a fan but if we have to keep it I would provide a lower bound like we do in requirements.txt

jannisborn replied:
not a fan at all either. If you have a better solution, I'd be more than happy to implement it.

jannisborn replied:
I set lower bounds to pytorch and pytorch-scatter now

drugilsberg commented on src/gt4sd/algorithms/generation/torchdrug/implementation.py:
```diff
@@ -0,0 +1,192 @@
+    hidden_dims: List[int] = [256, 256, 256, 256]
+
+    def __init__(self, resources_path: str):
+        """
```
add initial comment in docstring

jannisborn replied:
Done in [04fd6f6](https://github.com/GT4SD/gt4sd-core/pull/49/commits/04fd6f6d79c2703828c5c61c935673c2726eefad)

drugilsberg commented on src/gt4sd/frameworks/granular/arg_parser/parser.py:
```diff
@@ -94,3 +98,6 @@ def parse_arguments_from_config(conf_file: Optional[str] = None) -> argparse.Nam
     return result_namespace
+
+
+_sentencepiece
```
typo? I would remove this

jannisborn replied:
It's not a typo. This is needed because flake8 complains otherwise. The tests are executed with the `--disable-noqa` flag, so I dont see any other way of circumvening flake8 from complaining. The current configuration only ignores `__init__` files, see here: https://github.com/GT4SD/gt4sd-core/blob/d127590e2a312bda6c55f39304b70462f342ef4a/.github/workflows/tests.yaml#L45

If we import sentencepiece in the file, we have to use it somehow, otherwise we get a F401

jannisborn replied:
As I wrote above, this will not work because we run flake8 with `--disable-noqa` so we cannot overrule F401 errors, unless in `init` files.

drugilsberg replied:
I would honestly prefer to see _sentencepiece just below the import because it's a bit weird to see this there without comments.

drugilsberg commented on src/gt4sd/algorithms/generation/torchdrug/core.py:
```diff
@@ -0,0 +1,132 @@
+
+
+@ApplicationsRegistry.register_algorithm_application(TorchDrugGenerator)
+class TorchDrugZincGCPN(TorchDrugGCPNConfiguration):
```
Here I think we should have a unique configuration per network type.
No need to create one for each dataset. We can use the algorithm version to treat all datasets.
We can have as default `zinc250k` and then the others are simply other folders in the cache.
The `ApplicationRegistry` will take care of everything.

jannisborn replied:
Refactored this in [13a843f](https://github.com/GT4SD/gt4sd-core/pull/49/commits/13a843fd1622d5488eb76b895fd6a9501798ed2d), it's definitely more concise now

drugilsberg reviewed (changes requested):
Great job some general comments we need to address before merging:

- I would not have configuration classes for specific datasets. Since the classes are the same we should simply use versions for the different datatsets (see comment in PR).
- `sentencepiece` treatment. Instead of doing the trick every time I would do it once in  `src/gt4sd/__init__.py`. After the definition of `__version__` and `__name__` we can have the following:
```python
import sentencepiece as _sentencepiece  # noqa: F401
import pytorch_lightning as _pl  # noqa: F401
```
In this way we only have it there and inside the module we can import pytorch_lightning without any issue.
- I left a note on the installation.

Once those are addressed I would say we are good to go.

jannisborn commented:
I addressed all comments @drugilsberg, but unfortunately your solution with importing `sentencepiece` in the global init file does not work. Dependent on which functions we import first from the package, the `__init__` is not always executed before an import of lightning.

drugilsberg commented on src/gt4sd/algorithms/generation/torchdrug/abc.py:
```diff
@@ -0,0 +1,4 @@
+import torch
```
Do we still need this after the change in the setup? If yes, we should find to find a way to renable OpenMP after the import as quickly discussed.

jannisborn replied:
Yes, we still need this and this setup is indeed a way to renable openMP after the torchdrug import. 
We disable openMP in `abc.py` which always runs first in the init. In `core.py` we import the torchdrug things and then afterwards we disable it here again.

jannisborn replied:
Same problem like for the sentencepiece issue, if I put the the line to re-enable openmp in the torchdrug file, directly after the last torchdrug import, we get a E402 since we dont have all module imports at the top

jannisborn replied:
We could move it below the last import in the `implementation.py` but then I find it cleaner here especially since this init file is always executed even if we start a shell like:
```py
from gt4sd.algorithms.generation.torchdrug.implementation import *
```

drugilsberg replied:
makes sense 👍

drugilsberg replied:
Actually we could simply put a check and set it to False, and optionally reenable it, only in case had_openmp is indeed True at the time of import. Otherwise we risk to enable it even if OpenMP is not there.

jannisborn replied:
Done in 694b57f

jannisborn commented on src/gt4sd/algorithms/generation/torchdrug/core.py:
```diff
@@ -0,0 +1,113 @@
+            target: unused since it is not a conditional generator.
+
+        Example:
+            An example for using a generative algorithm from HuggingFace::
```
Good catch, resolved this. I did a grep, there's no other occurrences of this

drugilsberg reviewed (commented):
Looks good great job, just few comments remained.

drugilsberg commented on src/gt4sd/training_pipelines/pytorch_lightning/core.py:
```diff
@@ -4,13 +4,17 @@
 
 from ..core import TrainingPipeline, TrainingPipelineArguments
 
+# Sentencepiece has to be loaded before lightning
```
```suggestion
# sentencepiece has to be loaded before lightning to avoid segfaults
```

drugilsberg commented on src/gt4sd/training_pipelines/pytorch_lightning/language_modeling/core.py:
```diff
@@ -4,13 +4,17 @@
 from .lm_datasets import CGMDataModule, CLMDataModule, MLMDataModule, PLMDataModule
 from .models import LM_MODULE_FACTORY, CGMModule, CLMModule, MLMModule, PLMModule
 
+# Sentencepiece has to be loaded before lightning
```
```suggestion
# sentencepiece has to be loaded before lightning to avoid segfaults
```

drugilsberg commented on src/gt4sd/training_pipelines/pytorch_lightning/language_modeling/models.py:
```diff
@@ -19,6 +20,9 @@
     XLNetLMHeadModel,
 )
 
+# Sentencepiece has to be loaded before lightning
```
```suggestion
# sentencepiece has to be loaded before lightning to avoid segfaults
```

drugilsberg commented on src/gt4sd/frameworks/granular/arg_parser/parser.py:
```diff
@@ -2,11 +2,15 @@
 from ..ml.models import ARCHITECTURE_FACTORY
 from .utils import convert_string_to_class
 
+# Sentencepiece has to be loaded before lightning
```
```suggestion
# sentencepiece has to be loaded before lightning to avoid segfaults
```

drugilsberg commented on src/gt4sd/frameworks/granular/dataloader/data_module.py:
```diff
@@ -11,6 +12,9 @@
 from .dataset import CombinedGranularDataset, GranularDataset
 from .sampler import StratifiedSampler
 
+# Sentencepiece has to be loaded before lightning
```
```suggestion
# sentencepiece has to be loaded before lightning to avoid segfaults
```

drugilsberg commented on src/gt4sd/frameworks/granular/ml/module.py:
```diff
@@ -3,13 +3,17 @@
 from .models import GranularBaseModel, GranularEncoderDecoderModel
 from .models.model_builder import building_models, define_latent_models_input_size
 
+# Sentencepiece has to be loaded before lightning
```
```suggestion
# sentencepiece has to be loaded before lightning to avoid segfaults
```

drugilsberg commented on src/gt4sd/frameworks/granular/train/core.py:
```diff
@@ -14,6 +15,9 @@
 from ..ml.models import AUTOENCODER_ARCHITECTURES
 from ..ml.module import GranularModule
 
+# Sentencepiece has to be loaded before lightning
```
```suggestion
# sentencepiece has to be loaded before lightning to avoid segfaults
```

drugilsberg commented on src/gt4sd/algorithms/generation/torchdrug/abc.py:
```diff
@@ -0,0 +1,6 @@
+import torch
+
+# Disable openmp usage since this raises on MacOS when libomp has the wrong version.
```
```suggestion
# disable openmp usage since this raises on MacOS when libomp has the wrong version.
```

drugilsberg reviewed (approved):
Looks great well done
